Automattic/newspack-content-converter
# Add/retry failed conversions

Fixes #28 

**This PR brings two new features:**
- retrying conversion of just the posts which previously failed to get converted successfully,
- resetting an ongoing conversion, so that it can be completely restarted.

**TL;DR scenario** -- we'll create 300 test posts, that's 3 full batches, and rig the first half to fail conversion during the first run. Then we'll use the new functionality to retry converting just these previously failed Posts, and run in on those again. And we'll unrig them then, letting them pass this second time (this emulates a legitimate reason for failure now being cleared, such as backend being unresponsive before, but not any longer, etc). And we'll also test the resetting of an ongoing conversion, so that the conversion can be run all over again even though it was initialized and marked as "in progress".

**To test the PR:**
- generate test content:
    - first create exactly 300 public HTML Posts, and delete any other public Pages if there are any -- you can use the FakerPress Plugin to generate the 300 public HTML Posts
- install the plugin:
    - if the Plugin's already installed, first Deactivate & Delete it, then pull this branch and build it with `nvm use 10.10.0 && npm ci && npm run clean && npm run build:webpack`
    - additionally, if you're using a symlink to the Plugin's folder in your local env, and it may not be convenient for you to have to recreate the symlink after fully Deleting the plugin, feel free to pull, build and Activate according to these instructions https://github.com/Automattic/newspack-content-converter/pull/33#pullrequestreview-337412481 -- follow steps 1. and 2.
- purposely rig the first 150 Posts to fail during conversion (so that later on we have some failed ones to retry conversion):
    - run this SQL to set their content to invalid (one blank space causes a failure): `update ncc_wp_posts set post_content = '&nbsp;' order by ID limit 150;`
- run conversion for all content:
    - in WP-admin, go to "Newspack Content Converter" > "Run Conversion" > and click the big blue button "Start Conversion Now"
    - while conversion is running on the `/wp-admin/post-new.php?newspack-content-converter` page, open the Console, and observe that the first 150 posts fail with errors (you'll see output saying "A conversion error occured", and an error object with error message "No resulting blocks content") -- the whole first batch and a half of the second batch will fail, out of three batches total
    - after the 3 batches have completed with half of them failed, and half converted correctly, notice that the end message now says: "Warning: certain posts were not converted successfully..."
    - there click the "Back to Run Conversion page" link to `/wp-admin/admin.php?page=ncc-conversion` -- and then notice the new info section "Retry Converting Failed Posts" with another blue button "Retry Failed" -- don't click it yet
- now restore the content for the 150 failed posts, so that they will go through successfully the second time:
    - `update ncc_wp_posts nwp join wp_posts wp on wp.ID = nwp.ID set nwp.post_content = wp.post_content;`
- retry conversion only for the failed posts:
    - on the "Run Conversion" Page, click the "Retry Failed" second blue button
    - while conversion will now run 2 batches, in the Console observe that the 150 posts now get converted properly

If all this worked, it means that the new feature "retrying-conversion-of-only-failed-posts" works well :thumb_up.

**Let's also test the second feature added -- resetting an ongoing conversion:**
- start a conversion and purposely interrupt it:
    - in WP-admin, go to "Newspack Content Converter" > "Run Conversion" > and click the big blue button "Start Conversion Now"
    - once the conversion starts running on the `/wp-admin/post-new.php?newspack-content-converter` page, purposely close the active tab, and interrupt the conversion (another PR will add an extra warning to the user here, but there isn't any right now)
- reset ongoing conversion:
    - go to "Newspack Content Converter" > "Run Conversion" again, and notice the "Reset Ongoing Conversion in Order to Start Again" part
    - click the blue button "Reset Conversion"
    - confirm that the page now shows the "Start Conversion" button which lets you restart the conversion all over again, even though it's been closed unexpectedly -- if you wish, click it and confirm that the full 3 batches get converter successfully once again
